### ISSUE TITLE
narrow character + emoji variatiotion broken in vim again

### ISSUE THREAD

**Issue by ces42** [kovidgoyal/kitty#7329]
This is happening on e57692e4

**Describe the bug:**
write just "⚠️" into a text file. Then opening the file with vim (9.0) or neovim (lastest master) shows what looks like a file with just one space.

_Originally posted by @ces42 in https://github.com/kovidgoyal/kitty/issues/7005#issuecomment-1922745426_

**Comment by kovidgoyal** [comment]
This is a bug in those editors, you should report it to them. Proof:

The bug is not reproduced with

nvim --clean /t/emoji.txt

But it is reproduced with

nvim --clean -c 'set cursorline' /t/emoji.txt

The problem is the drawing of the cursorline highlight

In order to draw the cursor line highlight what nvim does is set the
style to underline, and then draw as many spaces as are left in the
line. The problem is that it ignores the variation width selector and
positions the cursor at the second cell of the emoji and then draws the
blanks. kitty now follows most terminal emulators and erases the emoji
if a character is drawn on its seconds cell. The bug is that nvim is
incorrectly calculating the width of the character, as 1 instead of 2.

You can see this happening by doing the following commnands and closing
the kitty window immediately after the screen is fully drawn

```
rm -r ~/.local/state/nvim/swap/ && kitty -o confirm_os_window_close=0 --dump-commands nvim --clean -c 'set cursorline' emoji.txt > bad
rm -r ~/.local/state/nvim/swap/ && kitty -o confirm_os_window_close=0 --dump-commands nvim --clean emoji.txt > good
```

Then diff good vs bad, you will see the difference is:

```
--- /t/good	2024-04-10 08:27:08.226332580 +0530
+++ /t/bad	2024-04-10 08:26:53.426210370 +0530
@@ -14,11 +14,15 @@ select_graphic_rendition 0
 screen_cursor_position 1 1
 screen_erase_in_display 2 0
 screen_push_key_encoding_flags 1
+screen_designate_charset 0 66
+select_graphic_rendition 0 4
 draw ⚠️
 screen_cursor_position 1 2
-screen_erase_in_line 0 0
+draw
 screen_carriage_return
 screen_linefeed
+screen_designate_charset 0 66
+select_graphic_rendition 0
 select_graphic_rendition 94
 draw ~
 screen_erase_in_line 0 0
 ```

In both cases nvim incorrectly moves the cursor to (y=1, x=2) aka the
second cell on the first line. It should be moving to (y=1, x=3). After
that in the good case it erases the rest of the line. kitty does not
erase an emoji in this case if the cursor is on the second cell. In the
bad case nvim is drawing spaces characters, in this case kitty does
erase the emoji because the second cell must contain a space character
which means the first cell cant have a width two emoji.

Bottom line, nvim needs to get the width calculation correct. And
incidentally, they should use character repeat instead of sending so
many repeated blanks. More efficient.

On a personal note, I avoid emoji with variation selectors for just this
reason. I use 🔴 🟡 🔵 🟣 instead for signs.
